include/deal.II/matrix_free/operators.h

```diff
@@ -1802,24 +1847,144 @@ namespace MatrixFreeOperators
+                                                              block_n),
+                                        diagonal_evaluation_f,
+                                        this->selected_rows[block_n]);
+
     inverse_diagonal_vector = diagonal_vector;
 
-    const unsigned int locally_owned_size =
-      inverse_diagonal_vector.locally_owned_size();
-    for (unsigned int i = 0; i < locally_owned_size; ++i)
-      inverse_diagonal_vector.local_element(i) =
-        Number(1.) / inverse_diagonal_vector.local_element(i);
+    for (unsigned int i = 0; i < inverse_diagonal_vector.locally_owned_size();
+         ++i)
+      if (std::abs(inverse_diagonal_vector.local_element(i)) >
+          std::sqrt(std::numeric_limits<Number>::epsilon()))
+        inverse_diagonal_vector.local_element(i) =
+          1. / inverse_diagonal_vector.local_element(i);
+      else
+        inverse_diagonal_vector.local_element(i) = 1.;
 
     inverse_diagonal_vector.update_ghost_values();
     diagonal_vector.update_ghost_values();
   }
 
 
 
+  template <int dim,
+            int fe_degree,
+            int n_q_points_1d,
+            int n_components,
+            typename VectorType,
+            typename VectorizedArrayType>
+  void
+  MassOperator<dim,
+               fe_degree,
+               n_q_points_1d,
+               n_components,
+               VectorType,
+               VectorizedArrayType>::compute_lumped_diagonal()
+  {
+    using Number =
+      typename Base<dim, VectorType, VectorizedArrayType>::value_type;
+    Assert((Base<dim, VectorType, VectorizedArrayType>::data.get() != nullptr),
+           ExcNotInitialized());
+    Assert(this->selected_rows == this->selected_columns,
+           ExcMessage("This function is only implemented for square (not "
+                      "rectangular) operators."));
+
+    inverse_lumped_diagonal_entries =
+      std::make_shared<DiagonalMatrix<VectorType>>();
+    lumped_diagonal_entries = std::make_shared<DiagonalMatrix<VectorType>>();
+    VectorType &inverse_lumped_diagonal_vector =
+      inverse_lumped_diagonal_entries->get_vector();
+    VectorType &lumped_diagonal_vector = lumped_diagonal_entries->get_vector();
+    this->initialize_dof_vector(inverse_lumped_diagonal_vector);
+    this->initialize_dof_vector(lumped_diagonal_vector);
+
+    // Re-use the inverse_lumped_diagonal_vector as the vector of 1s
+    inverse_lumped_diagonal_vector = Number(1.);
+    apply_add(lumped_diagonal_vector, inverse_lumped_diagonal_vector);
+    this->set_constrained_entries_to_one(lumped_diagonal_vector);
+
+    const size_type locally_owned_size =
+      inverse_lumped_diagonal_vector.locally_owned_size();
+    for (size_type i = 0; i < locally_owned_size; ++i)
+      inverse_lumped_diagonal_vector.local_element(i) =
+        Number(1.) / lumped_diagonal_vector.local_element(i);
+
+    inverse_lumped_diagonal_vector.update_ghost_values();
+    lumped_diagonal_vector.update_ghost_values();
```

Comment:
This makes me wonder whether we should really do this. We use the vectors primarily for linear algebra purposes, so I don't like to do this unconditionally as some functions like `Vector:add` then call a communication routine on every call, see https://github.com/dealii/dealii/blob/6dce9bff6e9a1526885ddb1738b02a6c25620b20/include/deal.II/lac/la_parallel_vector.templates.h#L1542-L1543

(There is nothing wrong here, but if others agree, I suggest we try to fix this up.)

Reply:
I saw that too - I don't think we ever need ghost values in this array, but I kept them since they were already present in the other `DiagonalMatrix`.